pkg/agent/protocol/http/proxy_test.go

```diff
@@ -371,35 +313,54 @@ func Test_ProxyHandler(t *testing.T) {
 		t.Run(tc.title, func(t *testing.T) {
 			t.Parallel()
 
-			client := &fakeHTTPClient{
-				body:    tc.body,
-				status:  tc.expectedStatus,
-				headers: tc.headers,
-			}
+			upstreamServer := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+				for k, values := range tc.headers {
+					for _, v := range values {
+						rw.Header().Add(k, v)
+					}
+				}
+				rw.WriteHeader(tc.statusCode)
 
-			upstreamURL, err := url.Parse(tc.config.UpstreamAddress)
+				_, err := rw.Write(tc.body)
+				if err != nil {
+					t.Errorf("writing upstream body: %v", err)
+				}
+			}))
+
+			upstreamURL, err := url.Parse(upstreamServer.URL)
 			if err != nil {
-				t.Errorf("error parsing upstream address %v", err)
-				return
+				t.Fatalf("error parsing httptest url")
 			}
+
 			handler := &httpHandler{
 				upstreamURL: *upstreamURL,
-				client:      client,
+				client:      http.DefaultClient,
 				disruption:  tc.disruption,
 				metrics:     &protocol.MetricMap{},
 			}
 
-			reqURL := fmt.Sprintf("http://%s%s", tc.config.ListenAddress, tc.path)
-			req := httptest.NewRequest(tc.method, reqURL, strings.NewReader(string(tc.body)))
-			recorder := httptest.NewRecorder()
-			handler.ServeHTTP(recorder, req)
-			resp := recorder.Result()
+			proxyServer := httptest.NewServer(handler)
+
+			req, err := http.NewRequest(tc.method, proxyServer.URL+tc.path, bytes.NewReader(tc.body))
+			if err != nil {
+				t.Fatalf("building request to proxy: %v", err)
+			}
+
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				t.Fatalf("making request to proxy: %v", err)
+			}
 
 			if tc.expectedStatus != resp.StatusCode {
 				t.Errorf("expected status code '%d' but '%d' received ", tc.expectedStatus, resp.StatusCode)
```

Comment:
As you are using `t.Fatal` above, shouldn't we use it uniformly at least in this test?